cmd2.py

```diff
@@ -969,25 +970,47 @@ def _surround_ansi_escapes(prompt, start="\x01", end="\x02"):
         return result
 
     def pseudo_raw_input(self, prompt):
-        """copied from cmd's cmdloop; like raw_input, but accounts for changed stdin, stdout"""
+        """
+        began life as a copy of cmd's cmdloop; like raw_input but
+        
+        - accounts for changed stdin, stdout
+        - if input is a pipe (instead of a tty), look at self.echo
+          to decide whether to print the prompt and the input
+        """
 
         # Deal with the vagaries of readline and ANSI escape codes
         safe_prompt = self._surround_ansi_escapes(prompt)
 
         if self.use_rawinput:
             try:
-                line = sm.input(safe_prompt)
+                if sys.stdin.isatty():
+                    sys.stdout.write(safe_prompt)
```

Comment:
Just curious ... why not just pass **safe_prompt** directly to ``sm.input(safe_prompt)`` like before?  Why call ``sys.stdout.write(safe_prompt)`` first and then ``sm.input()`` with no prompt?

Actually I did some manual testing, it needs to be the other way, very bad things can happen like the user can delete the prompt by backspacing far enough if you do it this way.